### PR TITLE
[Wip] Kubernetes talks to flannel-vxlan 

### PR DESCRIPTION
**Pull request by bprashanth**
Requires flannel as a godep with https://github.com/coreos/flannel/pull/308
Invoked via USE_OVERLAY="true" ./hack/local-up-cluster.sh or e2e.go -pushup

Only uploading so I can reference in networking discussions. Won't build because the godep isn't included, and including it will make the pr less referencable because of diff size. Godep commit hash 8b260bc6f04c322abd6483b477203c5db8c21e2a.

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
Will this ensure that the proper firewall ports are open on the hosts when it configures flannel for you?

**Comment by bprashanth** [comment]
Hmm.. firewall for the vxlan port? For external access or inside the cluster?

**Comment by derekwaynecarr** [comment]
@bprashanth - by default the port used by vxlan and flannel was blocked on my host and blocked pod-to-pod communication, so if kube is configuring things, it may want to open the firewall port on the host by doing the proper iptables rule.

**Comment by derekwaynecarr** [comment]
cc @kubernetes/rh-cluster-infra

**Comment by eparis** [comment]
@kubernetes/rh-networking

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -98,16 +98,13 @@ readonly KUBE_CLIENT_PLATFORMS=(
   linux/amd64
   linux/386
   linux/arm
-  darwin/amd64
-  darwin/386
-  windows/amd64
```
Comment (eparis) [inline hack/lib/golang.sh]: can't build on non-linux any more?

Comment (bprashanth) [inline hack/lib/golang.sh]: ah, heh, that's just a hack. Uh, one of the packages flannel includes (netlink i think) had documented problems building on non-linux. Punted on it for a while.

Comment (dcbw) [inline hack/lib/golang.sh]: > ah, heh, that's just a hack. One of the packages flannel includes (netlink i think) had documented problems building on non-linux. Punted on it for a while.

netlink is Linux specific.  And flannel uses netlink to create the vxlan interfaces at this point.  So yeah, anything using flannel + the VXLAN backend would be Linux-only.

**Comment by dcbw** [comment]
So is the point of this PR to implement the flannel server API in kubernetes so that the cluster doesn't have to run a master flanneld alongside the apiserver on the master?

**Comment by bprashanth** [comment]
Flannel doesn't have access to etcd, which is the important part. Kube pushed subnets down to the flannel daemon, not the other way aroud. No new resources are created to manage subnets etc (normal flanneld will create new kvs under --etcd-prefix subdir). 

This pr is just a proof of concept at the moment.

**Comment by bprashanth** [comment]
@wojtek-t rebased, latest patch set includes patched flannel godep

**Comment by dcbw** [comment]
IMO adding flannel bits into Kubernetes would be the wrong solution to the issue.  Instead, the solution would be more along the lines of:

1. Finish Kubernetes network plugin hooks to allow network providers (see https://github.com/kubernetes/kubernetes/pull/15465) which would add global hooks (instead of per-pod hooks like exist now with the CNI and 'exec' plugins)
2. Use those new hooks to implement your own plugin outside of Kubernetes that vendors flannel in and exposes the flannel API in the plugin, not in kubernetes
3. Profit!

We're doing almost exactly the same thing for openshift-sdn, except we'll be running a flannel master that talks to the kubernetes etcd instance, rather than stuffing all of flannel into the Kubernetes process itself.

**Comment by bprashanth** [comment]
Only reason it is the way it is, is because nothing else is ready yet. Even getting this for 1.2 and churning a little after that won't be too bad, it's isolated enough to split out eitherway. Exposing etcd to flannel probably won't fly, though.

**Comment by dcbw** [comment]
> Exposing etcd to flannel probably won't fly, though.

@bprashanth Why is that, especially if using flannel's listen/remote mode so that only the master flannel instance can talk to etcd?

**Comment by bprashanth** [comment]
@dcbw directly accessing etcd is verboten for a few reasons (api server checks schema of a database, the db is pluggable and can be mysql, we shard/replicate the apiserver not the db etc). It certainly is possible, and listen/remote makes it *less* egregious, but it will take some convincing before everyone just agress to writing to etcd. 

i.e we have a higher chance of getting stuff done by going through the apiserver, and an even higher chance if we just let the master assign the subnets like it does today (though I'll concede there are advantages to letting flannel do the latter, and that's an easier sell than the former).

**Comment by dcbw** [comment]
> i.e we have a higher chance of getting stuff done by going through the apiserver, and an even higher chance if we just let the master assign the subnets like it does today (though I'll concede there are advantages to letting flannel do the latter, and that's an easier sell than the former).

@bprashanth I see some of the arguments here; though personally (and I'm not authoritative on the matter) I really think anything like this, especially adding a whole new API surface, should be done in a network provider once the network provider proposals land.  That way the API is contained in your external network provider, and not added to Kubernetes itself.

**Comment by bprashanth** [comment]
We already hacked in flannel integration in for 1.2, hopefully it will get cleaned up through CNI plugins for 1.3

**Pull request closed without merging**